File: udf/agent/agent.go
package agent

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"sync"
)

// The Agent calls the appropriate methods on the Handler as it receives requests over a socket.
//
// Returning an error from any method will cause the Agent to stop and an ErrorResponse to be sent.
// Some *Response objects (like SnapshotResponse) allow for returning their own error within the object itself.
// These types of errors will not stop the Agent and Kapacitor will deal with them appropriately.
//
// The Handler is called from a single goroutine, meaning methods will not be called concurrently.
//
// To write Points/Batches back to the Agent/Kapacitor use the Agent.Responses channel.
type Handler interface {
	// Return the InfoResponse. Describing the properties of this Handler
	Info() (*InfoResponse, error)
	// Initialize the Handler with the provided options.
	Init(*InitRequest) (*InitResponse, error)
	// Create a snapshot of the running state of the handler.
	Snapshot() (*SnapshotResponse, error)
	// Restore a previous snapshot.
	Restore(*RestoreRequest) (*RestoreResponse, error)

	// A batch has begun.
	BeginBatch(*BeginBatch) error
	// A point has arrived.
	Point(*Point) error
	// The batch is complete.
	EndBatch(*EndBatch) error

	// Gracefully stop the Handler.
	// No other methods will be called.
	Stop()
}

// Go implementation of a Kapacitor UDF agent.
// This agent is responsible for reading and writing
// messages over a socket.
//
// The Agent requires a Handler object in order to fulfill requests.
type Agent struct {
	in  io.ReadCloser
	out io.WriteCloser

	outGroup     sync.WaitGroup
	outResponses chan *Response

	responses chan *Response
	// A channel for writing Responses, specifically Batch and Point responses.
	Responses chan<- *Response

	writeErrC chan error
	readErrC  chan error

	// The handler for requests.
	Handler Handler
}

// Create a new Agent is the provided in/out objects.
// To create an Agent that reads from STDIN/STDOUT of the process use New(os.Stdin, os.Stdout)
func New(in io.ReadCloser, out io.WriteCloser) *Agent {
	s := &Agent{
		in:           in,
		out:          out,
		outResponses: make(chan *Response),
		responses:    make(chan *Response),
	}
	s.Responses = s.responses
	return s
}

// Start the Agent, you must set an Handler on the agent before starting.
func (a *Agent) Start() error {
	if a.Handler == nil {
		return errors.New("must set a Handler on the agent before starting")
	}

	a.readErrC = make(chan error, 1)
	a.writeErrC = make(chan error, 1)
	a.outGroup.Add(1)
	go func() {
		defer a.outGroup.Done()
		err := a.readLoop()
		if err != nil {
			a.outResponses <- &Response{
				Message: &Response_Error{
					Error: &ErrorResponse{Error: err.Error()},
				},
			}
		}
		a.readErrC <- err
	}()
	go func() {
		a.writeErrC <- a.writeLoop()
	}()

	a.outGroup.Add(1)
	go func() {
		defer a.outGroup.Done()
		a.forwardResponses()
	}()

	return nil
}

// Wait for the Agent to terminate.
// The Agent will not terminate till the Responses channel is closed.
// You will need to close this channel externally, typically in the Stop method for the Handler.
// The Agent will terminate if the In reader is closed or an error occurs.
func (a *Agent) Wait() error {
	a.outGroup.Wait()
	close(a.outResponses)
	for a.readErrC != nil || a.writeErrC != nil {
		select {
		case err := <-a.readErrC:
			a.readErrC = nil
			if err != nil {
				return fmt.Errorf("read error: %s", err)
			}
		case err := <-a.writeErrC:
			a.writeErrC = nil
			if err != nil {
				return fmt.Errorf("write error: %s", err)
			}
		}
	}
	return nil
}

func (a *Agent) readLoop() error {
	defer a.Handler.Stop()
	defer a.in.Close()
	in := bufio.NewReader(a.in)
	var buf []byte
	request := &Request{}
	for {
		err := ReadMessage(&buf, in, request)
		if err == io.EOF {
			break
		}
		if err != nil {
			return err
		}

		// Hand message to handler
		var res *Response
		switch msg := request.Message.(type) {
		case *Request_Info:
			info, err := a.Handler.Info()
			if err != nil {
				return err
			}
			res = &Response{}
			res.Message = &Response_Info{
				Info: info,
			}
		case *Request_Init:
			init, err := a.Handler.Init(msg.Init)
			if err != nil {
				return err
			}
			res = &Response{}
			res.Message = &Response_Init{
				Init: init,
			}
		case *Request_Keepalive:
			res = &Response{
				Message: &Response_Keepalive{
					Keepalive: &KeepaliveResponse{
						Time: msg.Keepalive.Time,
					},
				},
			}
		case *Request_Snapshot:
			snapshot, err := a.Handler.Snapshot()
			if err != nil {
				return err
			}
			res = &Response{}
			res.Message = &Response_Snapshot{
				Snapshot: snapshot,
			}
		case *Request_Restore:
			restore, err := a.Handler.Restore(msg.Restore)
			if err != nil {
				return err
			}
			res = &Response{}
			res.Message = &Response_Restore{
				Restore: restore,
			}
		case *Request_Begin:
			err := a.Handler.BeginBatch(msg.Begin)
			if err != nil {
				return err
			}
		case *Request_Point:
			err := a.Handler.Point(msg.Point)
			if err != nil {
				return err
			}
		case *Request_End:
			err := a.Handler.EndBatch(msg.End)
			if err != nil {
				return err
			}
		}
		if res != nil {
			a.outResponses <- res
		}
	}
	return nil
}

func (a *Agent) writeLoop() error {
	defer a.out.Close()
	for response := range a.outResponses {
		err := WriteMessage(response, a.out)
		if err != nil {
			return err
		}
	}
	return nil
}

func (a *Agent) forwardResponses() {
	for r := range a.responses {
		a.outResponses <- r
	}
}

File: udf/agent/py/kapacitor/udf/agent.py
# Kapacitor UDF Agent implementation in Python
#
# Requires protobuf v3
#   pip install protobuf==3.11.1
from __future__ import absolute_import

import sys
from . import udf_pb2
from threading import Lock, Thread

try:
    from queue import Queue
except ImportError:
    from Queue import Queue


# Setup default in/out io
defaultIn = sys.stdin
defaultOut = sys.stdout

# Check for python3
# https://stackoverflow.com/a/38939320/703144
if sys.version_info >= (3, 0):
    defaultIn = sys.stdin.buffer
    defaultOut = sys.stdout.buffer

import io
import traceback
import socket
import os
import struct

import logging
logger = logging.getLogger()


# The Agent calls the appropriate methods on the Handler as requests are read off STDIN.
#
# Throwing an exception will cause the Agent to stop and an ErrorResponse to be sent.
# Some *Response objects (like SnapshotResponse) allow for returning their own error within the object itself.
# These types of errors will not stop the Agent and Kapacitor will deal with them appropriately.
#
# The Handler is called from a single thread, meaning methods will not be called concurrently.
#
# To write Points/Batches back to the Agent/Kapacitor use the Agent.write_response method, which is thread safe.
class Handler(object):
    def info(self):
        pass
    def init(self, init_req):
        pass
    def snapshot(self):
        pass
    def restore(self, restore_req):
        pass
    def begin_batch(self, begin_req):
        pass
    def point(self, point_req):
        pass
    def end_batch(self, end_req):
        pass



# Python implementation of a Kapacitor UDF agent.
# This agent is responsible for reading and writing
# messages over STDIN and STDOUT.
#
# The Agent requires a Handler object in order to fulfill requests.
class Agent(object):
    def __init__(self, _in=defaultIn, out=defaultOut,handler=None):
        self._in = _in
        self._out = out

        self._thread = None
        self.handler = handler
        self._write_lock = Lock()

    # Start the agent.
    # This method returns immediately
    def start(self):
        self._thread = Thread(target=self._read_loop)
        self._thread.start()

    # Wait for the Agent to terminate.
    # The Agent will terminate if STDIN is closed or an error occurs
    def wait(self):
        self._thread.join()
        self._in.close()
        self._out.close()

    # Write a response to STDOUT.
    # This method is thread safe.
    def write_response(self, response, flush=False):
        if response is None:
            raise Exception("cannot write None response")

        # Serialize message
        self._write_lock.acquire()
        try:
            data = response.SerializeToString()
            # Write message len
            encodeUvarint(self._out, len(data))
            # Write message
            self._out.write(data)
            if flush:
                self._out.flush()
        finally:
            self._write_lock.release()

    # Read requests off input stream
    def _read_loop(self):
        request = udf_pb2.Request()
        while True:
            msg = 'unknown'
            try:
                size = decodeUvarint32(self._in)
                data = self._in.read(size)

                request.ParseFromString(data)

                # use parsed message
                msg = request.WhichOneof("message")
                if msg == "info":
                    response = self.handler.info()
                    self.write_response(response, flush=True)
                elif msg == "init":
                    response = self.handler.init(request.init)
                    self.write_response(response, flush=True)
                elif msg == "keepalive":
                    response = udf_pb2.Response()
                    response.keepalive.time = request.keepalive.time
                    self.write_response(response, flush=True)
                elif msg == "snapshot":
                    response = self.handler.snapshot()
                    self.write_response(response, flush=True)
                elif msg == "restore":
                    response = self.handler.restore(request.restore)
                    self.write_response(response, flush=True)
                elif msg == "begin":
                    self.handler.begin_batch(request.begin)
                elif msg == "point":
                    self.handler.point(request.point)
                elif msg == "end":
                    self.handler.end_batch(request.end)
                else:
                    logger.error("received unhandled request %s", msg)
            except EOF:
                break
            except Exception as e:
                traceback.print_exc()
                error = "error processing request of type %s: %s" % (msg, e)
                logger.error(error)
                response = udf_pb2.Response()
                response.error.error = error
                self.write_response(response)
                break

# Indicates the end of a file/stream has been reached.
class EOF(Exception):
    pass

# Varint encode decode values
mask32uint = (1 << 32) - 1
byteSize = 8
shiftSize = byteSize - 1
varintMoreMask = 2**shiftSize
varintMask = varintMoreMask - 1


# Encode an unsigned varint
def encodeUvarint(writer, value):
    bits = value & varintMask
    value >>= shiftSize
    while value:
        writer.write(struct.pack("B", varintMoreMask | bits))
        bits = value & varintMask
        value >>= shiftSize
    return writer.write(struct.pack("B", bits))

# Decode an unsigned varint, max of 32 bits
def decodeUvarint32(reader):
    result = 0
    shift = 0
    while True:
        byte = reader.read(1)
        if len(byte) == 0:
            raise EOF
        b = struct.unpack("B", byte)[0]
        result |= ((b & varintMask) << shift)
        if not (b & varintMoreMask):
            result &= mask32uint
            return result
        shift += shiftSize
        if shift >= 32:
            raise Exception("too many bytes when decoding varint, larger than 32bit uint")

class Server(object):
    def __init__(self, socket_path, accepter):
        self._socket_path = socket_path
        self._listener = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM )
        self._listener.bind(socket_path)
        self._accepter = accepter

    def serve(self):
        self._listener.listen(5)
        try:
            while True:
                conn, addr = self._listener.accept()
                conn = conn.makefile(mode='rwb')
                thread = Thread(target=self._accepter.accept, args=(conn,addr))
                thread.start()
        except:
            self.stop()

    def stop(self):
        self._listener.close()
        try:
            os.remove(self._socket_path)
        except:
            pass

